Web Inspector Site Isolation Explainer v2 (#133)
WebKit/Documentation: docs/Deep Dive/Web Inspector/SiteIsolationExplainer.md

This design works perfectly when all frames share one process — but breaks down under Site
Isolation, where a `WebPageProxy` may have its frames distributed across several WebContent
Processes, each with its own `Page` and `PageInspectorController`.

---

## Background: The Inspector Target System

To persist a debugging session across WebProcess swaps (introduced with PSON), the concept of
_inspector targets_ was introduced. A target is an opaque handle that:

1. Provides a stable `targetId` the frontend can route commands to across process swaps.
2. Allows the same protocol interfaces to be reused across execution context types (Page,
   Worker, Frame).
3. Lets the frontend reason about the capabilities of each backend independently.

`WebPageInspectorController` in UIProcess manages the set of active targets. The `Target`
domain in `InspectorTargetAgent` exposes target lifecycle events (`Target.targetCreated`,
`Target.targetDestroyed`) to the frontend, and routes incoming commands to the correct target's
`BackendDispatcher`.

Before Site Isolation work, there were three target types:

- **`Page`** — legacy direct-backend target (pre-PSON and WebKitLegacy). No sub-targets.

> - **`WebPage`** — represents a `WebPageProxy`. May have transient worker and frame sub-targets.

Reviewer: ```suggestion
- **`WebPage`** — represents a `WebPageProxy`. May have transient worker sub-targets.
```

    *this, WebFrameInspectorTarget::toTargetID(frameID));
```

And in the destructor, the target is torn down symmetrically:

```cpp
page->inspectorController().destroyInspectorTarget(
    WebFrameInspectorTarget::toTargetID(frameID()));
```

This means frame targets are always present when frames exist, regardless of whether a
frontend is connected — consistent with how page and worker targets behave.

---

## Architecture: Target-Based Multiplexing

```
UIProcess
┌─────────────────────────────────────────────────────────┐
│  WebPageInspectorController                             │
│  ├── WebPageInspectorTargetProxy  (type: WebPage)       │
│  │     └── PageInspectorController  (in WCP-A)          │
│  ├── WebFrameInspectorTargetProxy  frame-1 (main)       │
│  │     └── FrameInspectorController  (in WCP-A)         │
│  └── WebFrameInspectorTargetProxy  frame-2 (cross-origin)
│        └── FrameInspectorController  (in WCP-B)         │
└─────────────────────────────────────────────────────────┘
         IPC ↕                    IPC ↕
  WebContent Process A      WebContent Process B
  PageInspectorController   (no PageInspectorController)

>   FrameInspectorController  FrameInspectorController

Reviewer: ```suggestion
         IPC ↕                    IPC ↕
  WebContent Process A      WebContent Process B
  PageInspectorController   PageInspectorController (not exposed)
  FrameInspectorController  FrameInspectorController
```
The `PageInspectorController` in process B still technically exists with a console agent created but is not used. (No other agents get created because that page's `connectFrontend` doesn't ever get called.) So, remote WebContent processes like process B rely on frame targets to make inspection work, rather than the unexposed page target

No frontend code needs to know whether it is talking to a single-process backend or a
Site-Isolated backend — the frame target abstraction provides uniform addressing.

---

## Remaining Work: Agent Migration Plan

Every inspector protocol domain must be adapted to work under Site Isolation. Each domain
falls into one of two migration patterns — **per-frame** or **octopus** — based on whether
its data is inherently scoped to a single frame or presents a unified page-level view.

### Per-Frame Domains

Per-frame domains get a new `Frame*Agent` subclass registered in
`FrameInspectorController::createLazyAgents()`. Commands route via the target system; events
flow through `FrontendRouter`. No new IPC is needed. IDs (`NodeId`, `StyleSheetId`, `ScriptId`,
etc.) are scoped per-target — the wire format doesn't change, but the frontend must pair each
ID with the target it came from.

| Domain | Status | Notes |
|--------|--------|-------|
| **Console** | Done | `FrameConsoleAgent` landed; reference implementation for all per-frame agents |
| **Runtime** | In progress | `FrameRuntimeAgent` (PR [#59021](https://github.com/WebKit/WebKit/pull/59021)) |

> | **Debugger** | Not started | Blocked on [bug 298909](https://bugs.webkit.org/show_bug.cgi?id=298909); single-debugger-multiple-agents challenge |

Reviewer 2: Does the “single-debugger-multiple-agents” problem have any known solution candidates? Or is it still in the exploratory phase? (What is the current status of bug 298909?)

Process captures instrumentation events and forwards them via IPC to a `Proxying*Agent`
aggregator in the UIProcess. The UIProcess agent handles command dispatch, ID remapping, and
presents a unified view to the frontend. A `Legacy*Agent` handles the WebKitLegacy
single-process path.

| Domain | Status | Notes |
|--------|--------|-------|
| **Network** | In progress | Core proxy/aggregator exists; commands (`getResponseBody`, etc.) still TODO |
| **Page** | In progress | Frame lifecycle events wired; `getResourceTree` aggregation still TODO |
| **Timeline** | Not started | Highest octopus priority; orchestrates sub-instruments |
| **ScriptProfiler** | Not started | Timeline sub-instrument |
| **Heap** | Not started | Large snapshot data; heap object ID remapping |
| **IndexedDB** | Not started | Origin-routed commands |
| **CPUProfiler** | Not started | `ENABLE(RESOURCE_USAGE)` only |
| **Memory** | Not started | `ENABLE(RESOURCE_USAGE)` only |
| **Audit** | Not started | May use target multiplexing instead of full octopus |

### UIProcess-Only

The **Browser** domain already lives entirely in `WebPageInspectorController` and requires
no migration.

> ### Migration Priority Order

Reviewer 2: Can we explain a bit about how the order is decided? (I guess it's user impact plus dependency)

1. **Foundation** (active): Console (done), Network, Page, Runtime
2. **Core frame inspection**: Debugger, DOM, CSS
3. **Dependent frame agents**: DOMDebugger, LayerTree, DOMStorage, Worker
4. **Remaining frame agents**: Canvas, Animation
5. **Page-level octopus agents**: Timeline, ScriptProfiler, Heap, IndexedDB, CPUProfiler, Memory, Audit

---

> ## Key Risks and Architectural Challenges

Author: Debugger is the main one where we have to make some decisions, but don't know which will work out yet. I added some notes.

potentially crossing frame boundaries.

### DOM: Cross-Frame Traversal at Process Boundaries

`InspectorDOMAgent` currently traverses into iframe `contentDocument` as children of
`HTMLFrameOwnerElement`. Under SI, this crosses process boundaries. `contentDocument` must be
omitted for out-of-process frames; the frontend discovers child frame DOM trees via frame
targets instead. `DOM.performSearch` changes from all-frames to per-frame scope.
`DOM.moveTo` across frames becomes invalid.

### CSS: Heavy DOM Agent Coupling

`InspectorCSSAgent` depends on `InspectorDOMAgent` for node resolution, document enumeration,
undo/redo history, and layout flag tracking. This coupling actually *simplifies* when both
agents are co-located per-frame (one document, one ID namespace, one history), but CSS
**must** migrate with or after DOM.

### Network: RequestId Collision Across Processes

`ResourceLoaderIdentifier` is generated per-process, so two WebContent Processes could
produce the same value. The UIProcess `ProxyingNetworkAgent` must qualify request IDs with the
source process to avoid incorrect resource lookups.

---

> ## Key Source Files

Author: I added some examples!

2. Allows the same protocol interfaces to be reused across execution context types (Page,
   Worker, Frame).
3. Lets the frontend reason about the capabilities of each backend independently.

`WebPageInspectorController` in UIProcess manages the set of active targets. The `Target`
domain in `InspectorTargetAgent` exposes target lifecycle events (`Target.targetCreated`,
`Target.targetDestroyed`) to the frontend, and routes incoming commands to the correct target's
`BackendDispatcher`.

Before Site Isolation work, there were three target types:

- **`Page`** — legacy direct-backend target (pre-PSON and WebKitLegacy). No sub-targets.
- **`WebPage`** — represents a `WebPageProxy`. May have transient worker and frame sub-targets.
- **`Worker`** — represents a dedicated Web Worker spawned from a Page.

Site Isolation adds a fourth:

- **`Frame`** — represents an individual `WebFrameProxy` / `LocalFrame`, each potentially in
  its own WebContent Process.

---

## Two Modes of Operation

### Mode 1: SI-disabled (or WebKitLegacy)

When Site Isolation is off, the architecture is essentially unchanged from the pre-SI model:

> - One `WebPageInspectorTargetProxy` (type `WebPage`) is created for the `WebPageProxy`.

Reviewer: ```suggestion
- One `PageInspectorTargetProxy` (type `Page`) is created for the `WebPageProxy`.
```
These `Web(Page|Frame)InspectorTarget(Proxy)?` things don't have the `Web` prefix now. (See, it even tripped you up! That target proxy is of type `Page`, not `WebPage`)

Reviewer: Actually, it might be more clear stated like this, highlighting the contrast between `WebPage` and `Page`:
```suggestion
- One `WebPageInspectorTargetProxy` (type `WebPage`) is created to manage the lifetime of the underlying `Page`, `Frame`, and `Worker` targets in the inspected webpage
- One `PageInspectorTargetProxy` (type `Page`) is created for the one `PageInspectorController`.
```

- All agents live in one `PageInspectorController` in one WebContent Process.
- `didCreateFrame` on `WebPageInspectorController` is a no-op — no frame targets are created.
- Commands are routed through the page target to `PageInspectorController`.

### Mode 2: SI-enabled

When Site Isolation is enabled, each `WebFrameProxy` gets its own inspector target:

- One `WebPageInspectorTargetProxy` still exists for page-level agents.
- Each `WebFrameProxy` creation triggers a `WebFrameInspectorTargetProxy` (type `Frame`).
- Each frame target connects to a `FrameInspectorController` in the owning WebContent Process.
- Commands targeted at a frame ID are routed to the correct `WebFrameInspectorTargetProxy`,
  which sends them over IPC to the `FrameInspectorController` in that process.

The key callsite is in `WebFrameProxy`'s constructor
(`UIProcess/WebFrameProxy.cpp`):

```cpp
page.inspectorController().createWebFrameInspectorTarget(

>     *this, WebFrameInspectorTarget::toTargetID(frameID));

Reviewer: ```suggestion
page.inspectorController().didCreateFrame(*this);
```
